### PR TITLE
OIDC SSO Release - 8 December 2023

### PR DESCRIPTION
**Pull request by MarkvanMents**

### CODE REVIEW

File: content/en/docs/appstore/modules/oidc.md [hunk 1]
```diff
@@ -95,6 +97,7 @@ It requires the following Marketplace modules to be included in your app:
 * [Community Commons](https://marketplace.mendix.com/link/component/170) – see [Community Commons](/appstore/modules/community-commons-function-library/) documentation
 * [Nanoflow Commons](https://marketplace.mendix.com/link/component/109515) – see [Nanoflow Commons](/appstore/modules/nanoflow-commons/) documentation
 * [Mx Model reflection](https://marketplace.mendix.com/link/component/69) – see [Mx Model Reflection](/appstore/modules/model-reflection/) documentation
+* [User Commons](/needalinkhere) (for version 2.4.0 and above)
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/oidc.md]: Is there already a link for the User Commons module? I can't find it in the Marketplace.

File: content/en/docs/appstore/modules/oidc.md [hunk 2]
```diff
@@ -433,12 +437,50 @@ Do not change the `UserProvisioning_StandardOIDC` microflow. This may give probl
...
+{{% /alert %}}
+
+You can set up custom user provisioning once your app is running using the `OIDC.OIDC_Client_Overview` page that you set up for the administrator for the app in [Configuring Navigation](#configure-nav). You can set up custom user provisioning as follows:
+
+1. Sign in to the running app with an administrator account.
+2. Navigate to the `OIDC.OIDC_Client_Overview` page set up in the app navigation.
+3. Select the **Provisioning** tab.
+4. Set up the following information:
+    
+    * **Custom user Entity (extension of System.User)** – the Mendix entity in which you will store and look up the user account. If you are using the Administration module this would be `Administration.Account`.
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/oidc.md]: What is an "extension of System.User". The terms we use in Mendix are "specialization" and "[generalization](https://docs.mendix.com/refguide/entities/#generalization)".

File: content/en/docs/appstore/modules/oidc.md [hunk 3]
```diff
@@ -433,12 +437,50 @@ Do not change the `UserProvisioning_StandardOIDC` microflow. This may give probl
...
+
+5. Under **Just in time provisioning**, for each piece of information you want to add to your custom user entity, select an **IdP Attribute** (claim) and specify the **Custom user Entity Attribute** where you want to store the information.
+
+    Note the following:
+
+    * You cannot use the IdP claim which is the primary attribute identifying the user and you cannot use the attribute you set in **The attribute where the user principal is stored**
+    * You can only map one IdP claim to a **Custom user Entity Attribute**
+    * IdP Attributes(Claims) cannot be of type enum, autonumber, or an association
+
+6. In **Select the CustomUserProvisioning**, select a microflow you want to run for [Custom User Provisioning Using a Microflow](#custom-provisioning-mf).
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/oidc.md]: In the paper document it says that this field is optional. How do you tell the module that you don't want to use a custom provisioning microflow?

I've assumed, below, that there will be a blank option which indicates this. Is this the case?

**Review by MarkvanMents — COMMENTED** [review]
I have a number of questions about this document.

File: content/en/docs/appstore/modules/oidc.md (hunk 2 above)
Comment (sowmyapathivada) [reply content/en/docs/appstore/modules/oidc.md]: CustomUserEntity (which has generalization with system.user)

File: content/en/docs/appstore/modules/oidc.md (hunk 3 above)
Comment (sowmyapathivada) [reply content/en/docs/appstore/modules/oidc.md]: yes Mark,the field is empty,If the user wants they can select the custom provisioning microflow

File: content/en/docs/appstore/modules/oidc.md (hunk 1 above)
Comment (sowmyapathivada) [reply content/en/docs/appstore/modules/oidc.md]: UserCommons is a new module that we are releasing in december

File: content/en/docs/appstore/modules/oidc.md (hunk 3 above)
Comment (MarkvanMents) [reply content/en/docs/appstore/modules/oidc.md]: Thanks.

File: content/en/docs/appstore/modules/oidc.md (hunk 2 above)
Comment (MarkvanMents) [reply content/en/docs/appstore/modules/oidc.md]: Will you change this in the UI at some stage - or have you. I took this text from the picture in the paper document.

File: content/en/docs/appstore/modules/oidc.md (hunk 1 above)
Comment (MarkvanMents) [reply content/en/docs/appstore/modules/oidc.md]: OK - I will need the link before I can publish this documentation.

File: content/en/docs/appstore/modules/oidc.md (hunk 2 above)
Comment (sowmyapathivada) [reply content/en/docs/appstore/modules/oidc.md]: I think ,this confirmation should be given by Jaap

Comment (MarkvanMents) [reply content/en/docs/appstore/modules/oidc.md]: This isn't necessary for this document. Resolving.

Comment (MarkvanMents) [reply content/en/docs/appstore/modules/oidc.md]: @jaapf
You may want to think about terminology in the next release as "extension" is not using the terms you find in the Studio Pro UX.

**Review by MarkvanMents — COMMENTED** [review]
Waiting for link to UserCommons, then this can be published.

File: content/en/docs/appstore/modules/oidc.md
Comment (sowmyapathivada) [inline content/en/docs/appstore/modules/oidc.md:1]: Document is impressively clear, concise, and visually appealing. Excellent work!

Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md:1]: Thank you @sowmyapathivada. Please have a look at other changes I made recently.

File: content/en/docs/appstore/modules/oidc.md [hunk 4]
```diff
@@ -43,13 +43,14 @@ Alternatives to using OIDC SSO for managing single sign-on are:
 
 * Supports SSO login with one or multiple OIDC/OAuth-compatible IdPs, such as AWS Cognito, Google, Salesforce, Apple, Okta, Ping, Microsoft's Entra ID (formerly known as Azure AD), and SAP Cloud Identity Services.
 * Easy configuration, by leveraging the so-called well-known discovery endpoint at your IdP.
-  * For example, PKCE will be used automatically if it is detected.
```
Comment (JaapF) [inline content/en/docs/appstore/modules/oidc.md]: An IdP's well-known endpoint also indicates which user claims the IdP may provide during SSO.  The module reads this information so the developer doesn't have to configure these. The available claims can be used in (custom) provisioning microflow, as decsribed in section x.y.z.

File: content/en/docs/appstore/modules/oidc.md [hunk 5]
```diff
@@ -305,7 +308,7 @@ In this case, the OIDC client is the app you are making.
...
     * If you need refresh tokens for your end-users, you also need the `offline_access` scope.
     * Add other scopes as needed.
-1. Select your user provisioning flow. By default, this module will use standard OpenID claims to provision end-users in your app. Also included is a flow that uses the standard UserInfo endpoint in OIDC, which is useful in the case that your IdP uses "thin tokens". Also included is a salesforce-specific example. You may need to make changes in order to implement provisioning logic which fits your business needs. To do so, read the section on [Custom User Provisioning](#custom-provisioning).
+1. Select your user provisioning flow. By default, this module will use standard OpenID claims to provision end-users in your app. Also included is a flow that uses the standard UserInfo endpoint in OIDC, which is useful in the case that your IdP uses "thin tokens". You can also use your own custom user entity to manage users of the app. See the section on [Custom User Provisioning](#custom-provisioning) for more information on what you can do to implement provisioning logic which fits your business needs. The module includes a Salesforce-specific example.
```
Comment (JaapF) [inline content/en/docs/appstore/modules/oidc.md]: I would expect a description of each of the (non-custom) user provsiioning flows that are available with the module. Which flow to use in which scenario.

File: content/en/docs/appstore/modules/oidc.md [hunk 6]
```diff
@@ -414,8 +417,9 @@ The following constants are optional:
 
 ## 6 User Provisioning
 
-Initially your app will not have any end-users. The OIDC module provides so-called Just-In-Time (JIT) user provisioning. This means that an end-user will be created in your app (as an `Account` object in the Administration module) when they log in for the first time.
-The user provisioning works by default but you can customize it if needed.
+Initially your app will not have any end-users. The OIDC module provides so-called Just-In-Time (JIT) user provisioning. This means that an end-user will be created in your app when they log in for the first time.
```
Comment (JaapF) [inline content/en/docs/appstore/modules/oidc.md]: If you don't want JIT user provisioning, it is possible to disable JIT, as described in configuraytion section x.y.z.

Comment (JaapF) [reply content/en/docs/appstore/modules/oidc.md]: Maybe add a reference to the next section about configuration of user provsiioning?

File: content/en/docs/appstore/modules/oidc.md [hunk 7]
```diff
@@ -43,13 +43,14 @@ Alternatives to using OIDC SSO for managing single sign-on are:
...
 * Easy configuration, by leveraging the so-called well-known discovery endpoint at your IdP.
-  * For example, PKCE will be used automatically if it is detected.
-* Configuration can be controlled through constants set during your deployment (version 2.3.0 and above).
-* Helper microflows (DELETE, GET, PATCH, POST, and PUT) which call an API with a valid token (and automate the token refresh process).
+    * For example, PKCE will be used automatically if it is detected.
+* Configuration can be controlled through constants set during your deployment (version 2.3.0 and above). 
+* Comes with helper microflows (DELETE, GET, PATCH, POST, and PUT) which call an API with a valid token (and automate the token refresh process).
 * Supports multiple OIDC IdPs by allowing configuration of user provisioning and access token parsing microflows per IdP.
 * Supports Authentication Context Class Reference (ACR) to allow your app to suggest the desired method or level of authentication for user login to the Identity Provider (IdP) (version 2.3.0 and above).
 * Supports responsive web applications, a.k.a. browser based applications.
```
Comment (JaapF) [inline content/en/docs/appstore/modules/oidc.md]: Harshita has been working on Progressive Web apps as well.
Please check with her and/or Phani.
We need to mention the PWA capability here and describe the specifics (if any?) for that as well.

File: content/en/docs/appstore/modules/oidc.md (hunk 5 above)
Comment (JaapF) [reply content/en/docs/appstore/modules/oidc.md]: We are changing the main identifier for users:
- so far we've been using "sub".
- starting with the next release we are switching to "prefered username".
I think this is characteristic of one of the user provisioning microflows.
Phani should help you with details.

File: content/en/docs/appstore/modules/oidc.md [hunk 8]
```diff
@@ -43,13 +43,14 @@ Alternatives to using OIDC SSO for managing single sign-on are:
 
 * Supports SSO login with one or multiple OIDC/OAuth-compatible IdPs, such as AWS Cognito, Google, Salesforce, Apple, Okta, Ping, Microsoft's Entra ID (formerly known as Azure AD), and SAP Cloud Identity Services.
```
Comment (JaapF) [inline content/en/docs/appstore/modules/oidc.md:48]: More new features:

* comes with default user provisioning microflow that works with Entra ID; there may need be a need to build a custom user provisioning flow
* From version x.y.z onwards, the module is compatible with the (forthcoming) SCIM module. For this is it is necessary to use the default microflow named XX.YY.ZZ.  This new workflow can only be used in any new application that you're building.

File: content/en/docs/appstore/modules/oidc.md [hunk 9]
```diff
@@ -95,6 +96,7 @@ It requires the following Marketplace modules to be included in your app:
 * [Community Commons](https://marketplace.mendix.com/link/component/170) – see [Community Commons](/appstore/modules/community-commons-function-library/) documentation
 * [Nanoflow Commons](https://marketplace.mendix.com/link/component/109515) – see [Nanoflow Commons](/appstore/modules/nanoflow-commons/) documentation
 * [Mx Model reflection](https://marketplace.mendix.com/link/component/69) – see [Mx Model Reflection](/appstore/modules/model-reflection/) documentation
+* [User Commons](/needalinkhere) (for version 2.4.0 and above)
```
Comment (JaapF) [inline content/en/docs/appstore/modules/oidc.md]: We need to add somthjing like.

- Existing apps cannot use the new microflow XX.YY.ZZ, which is necessary for using the forthcoming SCIM module.  This microflow is not backward compatible with previous versions of the OIDC module: previous versions used 'sub' as primary identifier, the microflow uses preferred_username.

File: content/en/docs/appstore/modules/oidc.md [hunk 10]
```diff
@@ -43,13 +43,14 @@ Alternatives to using OIDC SSO for managing single sign-on are:
...
 * Easy configuration, by leveraging the so-called well-known discovery endpoint at your IdP.
-  * For example, PKCE will be used automatically if it is detected.
-* Configuration can be controlled through constants set during your deployment (version 2.3.0 and above).
-* Helper microflows (DELETE, GET, PATCH, POST, and PUT) which call an API with a valid token (and automate the token refresh process).
+    * For example, PKCE will be used automatically if it is detected.
+* Configuration can be controlled through constants set during your deployment (version 2.3.0 and above). 
+* Comes with helper microflows (DELETE, GET, PATCH, POST, and PUT) which call an API with a valid token (and automate the token refresh process).
 * Supports multiple OIDC IdPs by allowing configuration of user provisioning and access token parsing microflows per IdP.
 * Supports Authentication Context Class Reference (ACR) to allow your app to suggest the desired method or level of authentication for user login to the Identity Provider (IdP) (version 2.3.0 and above).
```
Comment (JaapF) [inline content/en/docs/appstore/modules/oidc.md:54]: Can i suggest we split up the list of features into three parts:
IdP Integration capabilities
configuration experince features
developer experience features

File: content/en/docs/appstore/modules/oidc.md (hunk 5 above)
Comment (JaapF) [reply content/en/docs/appstore/modules/oidc.md]: One of the default user provisioning flows sets the timezone and language. Let's explain which one, so the developer can select that one if he needs that.
Phani knows the details, we discussed yesterday.

File: content/en/docs/appstore/modules/oidc.md [hunk 11]
```diff
@@ -43,13 +43,14 @@ Alternatives to using OIDC SSO for managing single sign-on are:
 
 * Supports SSO login with one or multiple OIDC/OAuth-compatible IdPs, such as AWS Cognito, Google, Salesforce, Apple, Okta, Ping, Microsoft's Entra ID (formerly known as Azure AD), and SAP Cloud Identity Services.
 * Easy configuration, by leveraging the so-called well-known discovery endpoint at your IdP.
-  * For example, PKCE will be used automatically if it is detected.
-* Configuration can be controlled through constants set during your deployment (version 2.3.0 and above).
-* Helper microflows (DELETE, GET, PATCH, POST, and PUT) which call an API with a valid token (and automate the token refresh process).
+    * For example, PKCE will be used automatically if it is detected.
+* Configuration can be controlled through constants set during your deployment (version 2.3.0 and above). 
+* Comes with helper microflows (DELETE, GET, PATCH, POST, and PUT) which call an API with a valid token (and automate the token refresh process).
 * Supports multiple OIDC IdPs by allowing configuration of user provisioning and access token parsing microflows per IdP.
```
Comment (JaapF) [inline content/en/docs/appstore/modules/oidc.md:53]: User provisioning microflows can be anywhere in your app; the do not need to be part of the OIDC module.

File: content/en/docs/appstore/modules/oidc.md (hunk 8 above)
Comment (JaapF) [reply content/en/docs/appstore/modules/oidc.md:48]: We need to be clear what is supported in which module versions and which module versions can be used with Mx9 and/or Mx10.

File: content/en/docs/appstore/modules/oidc.md [hunk 12]
```diff
@@ -43,13 +43,14 @@ Alternatives to using OIDC SSO for managing single sign-on are:
...
 * Supports Authentication Context Class Reference (ACR) to allow your app to suggest the desired method or level of authentication for user login to the Identity Provider (IdP) (version 2.3.0 and above).
 * Supports responsive web applications, a.k.a. browser based applications.
 * Works with the Mendix DeepLink module.
+* Supports user provisioning to custom user entities; you can map claims onto attributes of an entity which is a specialization of the `System.User` entity.
 * Built primarily in standard Mendix components (minimal Java) to allow for easy customization and ongoing development.
```
Comment (JaapF) [inline content/en/docs/appstore/modules/oidc.md:70]: Let's put an extra sub-heading here - "OIDC protocol adherence" - we also have that with SAML and SCIM.

File: content/en/docs/appstore/modules/oidc.md (hunk 7 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md]: Mentioned PWA in this commit: https://github.com/mendix/docs/pull/6922/commits/298bbbbac9dd699e0ee561aa2e7c9e69807fec82

File: content/en/docs/appstore/modules/oidc.md [hunk 13]
```diff
@@ -308,7 +310,17 @@ In this case, the OIDC client is the app you are making.
...
     * If you need refresh tokens for your end-users, you also need the `offline_access` scope.
     * Add other scopes as needed.
-1. Select your user provisioning flow. By default, this module will use standard OpenID claims to provision end-users in your app. Also included is a flow that uses the standard UserInfo endpoint in OIDC, which is useful in the case that your IdP uses "thin tokens". You can also use your own custom user entity to manage users of the app. See the section on [Custom User Provisioning](#custom-provisioning) for more information on what you can do to implement provisioning logic which fits your business needs. The module includes a Salesforce-specific example.
+1. Select your user provisioning flow. By default, this module will use standard OpenID claims to provision end-users in your app. Also included is a flow that uses the standard UserInfo endpoint in OIDC, which is useful in the case that your IdP uses "thin tokens". You can set up user provisioning by setting the following standard flows:
+
+| Default Microflow | Use |
+| --- | --- |
+| OIDC_CustomUserParsing_Standard | xxxxxxxxxx |
```
Comment (Karuna-Mendix) [inline content/en/docs/appstore/modules/oidc.md]: @sowmyapathivada, Please provide me the uses of these microflows

File: content/en/docs/appstore/modules/oidc.md [hunk 14]
```diff
@@ -308,7 +310,17 @@ In this case, the OIDC client is the app you are making.
...
+| OIDC_CustomUserParsing_UserInfo | xxxxxx |
+| OIDC_CustomUserParsing_Salesforce | xxxxxxxxxx |
+
+You can set the timezone and language with `xxxxx` or `xxxxxx` microflow.
```
Comment (Karuna-Mendix) [inline content/en/docs/appstore/modules/oidc.md]: @sowmyapathivada Please provide me the names of these microflows supporting timezone and language.

File: content/en/docs/appstore/modules/oidc.md (hunk 4 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md]: https://github.com/mendix/docs/pull/6922/commits/8cea9079a57604b7311e7b2bd176ac3774954e5e

File: content/en/docs/appstore/modules/oidc.md (hunk 11 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md:53]: https://github.com/mendix/docs/pull/6922/commits/8cea9079a57604b7311e7b2bd176ac3774954e5e

File: content/en/docs/appstore/modules/oidc.md (hunk 6 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md]: https://github.com/mendix/docs/pull/6922/commits/8cea9079a57604b7311e7b2bd176ac3774954e5e

File: content/en/docs/appstore/modules/oidc.md (hunk 5 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md]: https://github.com/mendix/docs/pull/6922/commits/8cea9079a57604b7311e7b2bd176ac3774954e5e

File: content/en/docs/appstore/modules/oidc.md (hunk 8 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md:48]: SCIM specific info will be published at a time of SCIM release.

Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md:48]: https://github.com/mendix/docs/pull/6922/commits/333ad3126ee9166999ff4fe50796262740a92546

File: content/en/docs/appstore/modules/oidc.md (hunk 10 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md:54]: https://github.com/mendix/docs/pull/6922/commits/2c59b17d58fad3bc243611e18458cd3ec2d1d52a

File: content/en/docs/appstore/modules/oidc.md (hunk 9 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md]: https://github.com/mendix/docs/pull/6922/commits/80c7af2231b552ef14da9b89da13f27d9f33ad22

File: content/en/docs/appstore/modules/oidc.md (hunk 12 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md:70]: https://github.com/mendix/docs/pull/6922/commits/2c59b17d58fad3bc243611e18458cd3ec2d1d52a

File: content/en/docs/appstore/modules/oidc.md (hunk 14 above)
Comment (Karuna-Mendix) [reply content/en/docs/appstore/modules/oidc.md]: https://github.com/mendix/docs/pull/6922/commits/2c59b17d58fad3bc243611e18458cd3ec2d1d52a

**Pull request merged**